scipy/optimize/_numdiff.py

```diff
@@ -579,80 +652,117 @@ def matvec(p):
     else:
         raise RuntimeError("Never be here.")
 
-    return LinearOperator((m, n), matvec)
+    return LinearOperator((m, n), matvec), nfev
 
 
-def _dense_difference(fun, x0, f0, h, use_one_sided, method):
+def _dense_difference(fun, x0, f0, h, use_one_sided, method, workers):
     m = f0.size
     n = x0.size
     J_transposed = np.empty((n, m))
-    x1 = x0.copy()
-    x2 = x0.copy()
-    xc = x0.astype(complex, copy=True)
+    nfev = 0
 
-    for i in range(h.size):
-        if method == '2-point':
-            x1[i] += h[i]
-            dx = x1[i] - x0[i]  # Recompute dx as exactly representable number.
-            df = fun(x1) - f0
-        elif method == '3-point' and use_one_sided[i]:
-            x1[i] += h[i]
-            x2[i] += 2 * h[i]
-            dx = x2[i] - x0[i]
-            f1 = fun(x1)
-            f2 = fun(x2)
-            df = -3.0 * f0 + 4 * f1 - f2
-        elif method == '3-point' and not use_one_sided[i]:
-            x1[i] -= h[i]
-            x2[i] += h[i]
-            dx = x2[i] - x1[i]
-            f1 = fun(x1)
-            f2 = fun(x2)
-            df = f2 - f1
-        elif method == 'cs':
-            xc[i] += h[i] * 1.j
-            f1 = fun(xc)
-            df = f1.imag
-            dx = h[i]
-        else:
-            raise RuntimeError("Never be here.")
+    if method == '2-point':
+        def x_generator2(x0, h):
+            for i in range(n):
+                # If copying isn't done then it's possible for different workers
+                # to see the same values of x1. (At least that's what happened
+                # when I used `multiprocessing.dummy.Pool`).
+                # I also considered creating all the vectors at once, but that
+                # means assembling a very large N x N array. It's therefore a
+                # trade-off between N array copies or creating an NxN array.
+                x1 = np.copy(x0)
+                x1[i] = x0[i] + h[i]
+                yield x1
+
+        f_evals = workers(fun, x_generator2(x0, h))
+        dx = [(x0[i] + h[i]) - x0[i] for i in range(n)]
+        df = [f_eval - f0 for f_eval in f_evals]
+        df_dx = [delf / delx for delf, delx in zip(df, dx)]
+        nfev += len(df_dx)
```

Comment:
Optional: I suppose there's a reason you're splitting the computations like this; from just reading the code it's not clear why numerators go through `workers` but not denominators. Consider adding a code comment

Reply:
numerators are where the objective function gets evaluated (expensive). The denominator is a simple array subtraction/addition.

Reply:
Well, that I understand. My only point here is that the earlier code was easy to understand, the current one is trickier, and the split between `delf` and `delx` is not helping. If possible, I'd prefer to read code where they are close to each other. And since denominators are indeed cheap, maybe there's a simple way to have them computed closer to where the `delf` is.

Reply:
Would you prefer that 678, 679,680 are evaluated in a single comprehensiom?

Reply:
As a mild preference, yes.
```
df_dx = [(f_eval[i] - f0) / ((x0[i] + h[i]) - x0[i]) for i in range(n)]
```

However if the current version is more readable to you, fine by me.

Reply:
>  is more readable to you

I find find it easier to read the way it is at the moment.